src/main/java/org/spongepowered/common/registry/TagRegistry.java

```diff
@@ -0,0 +1,184 @@
+    }
+
+    @Nullable
+    @Override
+    public ResourceLocation getKey(final Tag<T> var1) {
+        return this.staticTagHelper.getAllTags().getId((net.minecraft.tags.Tag<T>) var1);
+    }
+
+    @Override
+    public Optional<net.minecraft.resources.ResourceKey<Tag<T>>> getResourceKey(Tag<T> var1) {
+        final ResourceLocation valueLocation = this.getKey(var1);
+        if (valueLocation == null) {
+            return Optional.empty();
+        }
+        return Optional.of(net.minecraft.resources.ResourceKey.create(this.key(), valueLocation));
+    }
+
+    @Override
+    public int getId(@Nullable final Tag<T> var1) {
+        throw new UnsupportedOperationException("No ids not supported in TagRegistry!");
+    }
+
+    @Nullable
+    @Override
+    public Tag<T> byId(final int var1) {
+        throw new UnsupportedOperationException("No ids not supported in TagRegistry!");
+    }
+
+    @Nullable
+    @Override
+    public Tag<T> get(@Nullable final net.minecraft.resources.ResourceKey<Tag<T>> var1) {
+        if (!var1.isFor(this.key())) {
+            throw new IllegalStateException("Minecraft ResourceKey " + var1 + " is not for registry " + this.key());
+        }
+        return this.get(var1.location());
+    }
+
+    @Nullable
+    @Override
+    public Tag<T> get(@Nullable final ResourceLocation key) {
+        return this.findValue((ResourceKey) (Object) key).orElse(null);
+    }
+
+    @Override
+    protected Lifecycle lifecycle(Tag<T> var1) {
+        return this.lifecycle;
+    }
+
+    @Override
+    public Lifecycle elementsLifecycle() {
+        return this.lifecycle;
+    }
+
+    @Override
+    public Set<ResourceLocation> keySet() {
+        return Collections.unmodifiableSet(this.staticTagHelper.getAllTags().getAllTags().keySet());
+    }
+
+    @Override
+    public Set<Map.Entry<net.minecraft.resources.ResourceKey<Tag<T>>, Tag<T>>> entrySet() {
+        return Collections.unmodifiableSet(this.staticTagHelper.getAllTags().getAllTags().entrySet().stream()
+                .map(entry -> new AbstractMap.SimpleEntry<>(
+                        net.minecraft.resources.ResourceKey.create(this.key(), entry.getKey()),
+                        (Tag<T>) entry.getValue()))
+                .collect(Collectors.toSet()));
+    }
+
+    @Override
+    public Stream<Tag<T>> stream() {
+        return this.staticTagHelper.getWrappers().stream().map(tag -> (Tag<T>) tag);
+    }
+
+    @Override
+    public boolean containsKey(ResourceLocation var1) {
+        return this.staticTagHelper.getAllTags().getAllTags().containsKey(var1);
+    }
+
+    @Override
+    public boolean isDynamic() {
+        // Its dynamic, but this is not the place to register.
+        // I think its best to return true, because its possible a plugin
+        // would use this to determine whether it can be cached, and it can't.
+        return true;
```

Comment:
I would go for `ReadOnlyRegistry` myself since any future writes to it are literally it being overwritten.

Additionally dual is correct, no registry value can be cached. It is not legal per API spec. I suspect people will do it anyways but we shouldn't contribute confusion. This should return false.